lib/PreviewProvider.js

```diff
@@ -0,0 +1,239 @@
+/* globals Task */
+"use strict";
+
+const {Cu} = require("chrome");
+const ss = require("sdk/simple-storage");
+const {setTimeout, clearTimeout} = require("sdk/timers");
+const EMBEDLY_ENDPOINT = require("sdk/simple-prefs").prefs.embedlyEndpoint;
+
+const EMBEDLY_PROXY_MAX_LINKS = 25; // number of links embedly proxy accepts per request
+const CLEAR_CACHE_TIMEOUT = 86400000; // 24 hours to clear cache
+const REMOVE_KEYS = ["auth", "password", "username", "pass", "user"];
+const ALLOWED_PROTOCOLS = new Set(["http:", "https:"]);
+const DISALLOWED_HOSTS = new Set(["localhost", "127.0.0.1", "0.0.0.0"]);
+const URL_FILTERS = [
+  (item) => !!item.url,
+  (item) => !!(new URL(item.url)),
+  (item) => ALLOWED_PROTOCOLS.has(new URL(item.url).protocol),
+  (item) => !DISALLOWED_HOSTS.has(new URL(item.url).hostname)
+];
+
+Cu.importGlobalProperties(["fetch"]);
+Cu.importGlobalProperties(["URL"]);
+Cu.import("resource://gre/modules/Task.jsm");
+
+function PreviewProvider() {
+  this.init();
+}
+
+PreviewProvider.prototype = {
+  _timeoutID: null,
+
+  /**
+    * Clean up cache on addon bootup
+    */
+  cleanUpCache: Task.async(function*() {
```

Comment:
You could check if a key's timeout has expired at key lookup time rather than setting up a timeout, and then you don't need a background process at all.